### PR TITLE
Add optimized container for when alive children are much fewer than all children

### PR DESCRIPTION
**Pull request by ekrctb**
This PR adds `LifetimeManagementContainer` <sup>[1](#fn1)</sup>.
The container should behave the same as normal `CompositeDrawable` <sup>[2](#fn2)</sup> but optimized when number of alive children is significantly smaller than number of all children.
<a name="fn1">1</a>: I'm not sure about this name, suggestions are welcome.
<a name="fn2">2</a>: As long as assumptions are satisfied.

The container can be used as a base class of osu! `HitObjectContainer`, `DrawableStoryboardLayer` and `ConnectionRenderer` to improve performance of long and/or storyboard-heavy maps. I measured for a map and the result is from ~250fps to ~350fps.
Edit: By also adopting for approach circle proxies, it achieved 1000+fps!

### CODE REVIEW

File: osu.Framework/Graphics/Containers/CompositeDrawable.cs [hunk 1]
```diff
@@ -565,7 +565,7 @@ protected virtual bool UpdateChildrenLife()
             if (LoadState < LoadState.Ready)
                 return false;
 
-            if (!checkChildrenLife())
+            if (!CheckChildrenLife())
```
Comment (UselessToucan) [inline osu.Framework/Graphics/Containers/CompositeDrawable.cs]: Can be shortened to `return CheckChildrenLife()`

File: osu.Framework/Graphics/Containers/LifetimeManagementContainer.cs [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+            {
+                // Largest LifetimeEnd is placed on last.
+                return (x?.LifetimeEnd ?? 0).CompareTo(y?.LifetimeEnd ?? 0);
+            }
+        }
+
+        // Not yet loaded children.
+        private readonly List<Drawable> newChildren = new List<Drawable>();
+
+        // Children which is currently dead and becomes alive in future (in the Clock sense).
```
Comment (UselessToucan) [inline osu.Framework/Graphics/Containers/LifetimeManagementContainer.cs]: Would prefer multiline or xmldoc

Comment (ekrctb) [reply osu.Framework/Graphics/Containers/LifetimeManagementContainer.cs]: Normal comment because these are private stuff

Comment (peppy) [reply osu.Framework/Graphics/Containers/LifetimeManagementContainer.cs]: I'd still use xmldoc for private stuff, helps with intellisense

File: osu.Framework/Graphics/Containers/LifetimeManagementContainer.cs [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+        // Child with earliest LifetimeStart is placed at the last in order to do RemoveAt(Lengh - 1) in O(1) time rather than RemoveAt(0) in O(length) time.
+        private readonly SortedList<Drawable> nextChildren = new SortedList<Drawable>(new LifetimeStartComparator());
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/LifetimeManagementContainer.cs]: Let's call this `futureChildren`

File: osu.Framework/Graphics/Containers/LifetimeManagementContainer.cs [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+        /// Children which is currently dead and becomes alive if the clock is rewinded.
+        private readonly SortedList<Drawable> prevChildren = new SortedList<Drawable>(new LifetimeEndComparator());
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/LifetimeManagementContainer.cs]: `pastChildren`

File: osu.Framework/Graphics/Containers/LifetimeManagementContainer.cs [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+        // newChildren are not on this map.
+        private readonly Dictionary<Drawable, ChildState> childStateMap = new Dictionary<Drawable, ChildState>();
+
+        private enum ChildState
+        {
+            // This child is in AliveInternalChildren.
+            Alive,
+
+            // This child is in nextChildren.
+            Next,
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/LifetimeManagementContainer.cs]: `Future`

File: osu.Framework/Graphics/Containers/LifetimeManagementContainer.cs [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+
+            // This child is in prevChildren.
+            Prev,
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/LifetimeManagementContainer.cs]: `Past`

**Comment by peppy** [comment]
Discussing with @smoogipoo (in relation to his feedback on the osu-side PR), we can probably fix the remaining scenario by adding any children that bypass the alive state (`Past` -> `Future` or reverse) to a new list, which is processed once in `UpdateSubTree`, calling `child.UpdateSubTree()` once on the current frame then being cleared.

We also need support for changing `LifetimeStart`/`LifetimeEnd` (ie. for editor support where `DrawableHitObjects` are moved in time). Either making `LifetimeStart`/`LifetimeEnd`'s setters fire an (`internal`?) changed method, or making them `private set` and adding a new `ChangeLifetime(double start, double end)` pathway.

As a future consideration, we may want to remove `ShouldBeAlive`'s `virtual` keyword as it is used in few places.

File: osu.Framework/Graphics/Containers/LifetimeManagementContainer.cs [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+        /// Children which is currently dead and becomes alive if the clock is rewinded.
+        private readonly SortedList<Drawable> prevChildren = new SortedList<Drawable>(new LifetimeEndComparator());
+
+        // newChildren are not on this map.
+        private readonly Dictionary<Drawable, ChildState> childStateMap = new Dictionary<Drawable, ChildState>();
+
+        private enum ChildState
+        {
+            // This child is in AliveInternalChildren.
+            Alive,
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Containers/LifetimeManagementContainer.cs]: `Current` (keeps it in line with all other renames)

**Comment by ekrctb** [comment]
I now think rather than having `OnChildLifetimeSkipped` only on skipped children, probably useful and symmetric to have different kind of callbacks: when `LifetimeStart` is crossed by time delta (either (past time < LifetimeStart <= current time) or (current time < LifetimeStart <= past time)) and similar thing for `LifetimeEnd`. Skipping whole lifetime interval causes two callbacks in one frame.
If one want to ensure something is processed until lifetime is end then one processing can be done when `LifetimeEnd` is crossed forwardly.

**Comment by peppy** [comment]
Sure, sounds good. I don't like calling this "skipped" either. As long as this still fulfills the osu-side requirements.

**Comment by peppy** [comment]
Let's get this in for further testing on osu-side 👍

**Pull request merged**